XivAlexander/App_Feature_AnimationLockLatencyHandler.cpp

```diff
@@ -152,18 +154,73 @@ class App::Feature::AnimationLockLatencyHandler::Internals {
 										const int64_t rtt = now - m_latestSuccessfulRequest.RequestTimestamp;
 										conn.AddServerResponseDelayItem(rtt);
 
+										const int64_t rtt_med = conn.GetMedianServerResponseDelay();
+										const int64_t rtt_dev = conn.GetServerResponseDelayDeviation();
+
+										extraMessage = Utils::FormatString(" rtt=%lldms/%lldms/%lldms", rtt, rtt_med, rtt_dev);
+
 										m_latestSuccessfulRequest.ResponseTimestamp = now;
 
 										int64_t extraDelay = ExtraDelay;
-										const int64_t latency = conn.GetConnectionLatency();
-										const int64_t delay = conn.GetMedianServerResponseDelay();
+										int64_t delay = rtt;
+
 										if (config.UseAutoAdjustingExtraDelay) {
-											if (latency > 0 && delay > 0) {
-												extraDelay = std::max(0LL, delay - latency);
-												extraMessage = Utils::FormatString(" latency=%lldms delay=%lldms extraDelay=%lldms", latency, delay, extraDelay);
+											// Get current latency data
+											const int64_t latency_orig = conn.GetConnectionLatency();
+											int64_t latency = latency_orig;
+
+											extraMessage += Utils::FormatString(" latency=%lldms", latency);
+
+											// Update latency statistics
+											conn.AddConnectionLatencyItem(latency);
+
+											if (config.UseLatencyCorrection) {
+												/*
+												// Set server response delay to predicted value as pivot point.
+												delay = rtt_med;
+
+												if (rtt > rtt_med)
+													delay += rtt_dev;
+
+												if (rtt < rtt_med)
+													delay -= rtt_dev;
+												
+												// Use current server response delay to get a better prediction.
+												delay += rtt;
+												delay /= 2;
+												*/
+
+												// Get latency statistic data
+												const int64_t latency_med = conn.GetMedianConnectionLatency();
+												const int64_t latency_dev = conn.GetConnectionLatencyDeviation();
+
+												extraMessage += Utils::FormatString("/%lldms/%lldms", latency_med, latency_dev);
+
+												/*
+												// Adjust latency calculation with deviation and median.
+												latency = latency_med;
+
+												if (latency_orig > latency_med)
+													latency += latency_dev;
+
+												if (latency_orig < latency_med)
+													latency -= latency_dev;
+												*/
+
+												// Apply base latency with deviation.
+												// This is essentially a penalty for fluctuating connections. However, it will also help prevent overcompensating.
+												const int64_t latency_base = config.BaseLatencyPenalty;
+												int64_t penalty = (latency_med < latency_base) ? ((latency_med + latency_dev) / 2) : std::max(latency_base / 2, latency_dev);
+
+												latency = std::max(penalty, latency - penalty);
```

Comment:
It would be nice if we can get an extra explanation on the last lines, and remove the WIP? codes in comment blocks.

Reply:
Sure, I will add more comments here. The commented code is a leftover from testing calculations using different statistics. I will remove those as well.